# Human Activity Recognizer

[![Twitter Follow](https://img.shields.io/twitter/follow/dialhaseeb?style=social)](www.twitter.com/dialhaseeb)

![Logo](https://github.com/zenyc/zenyc/blob/master/logo-small.png)

## 🕯 About
**human-activty-recognizer** is a *machine learning model* that predicts human activity by using smartphone sensor data. It was trained using XGBoost (eXtreme Gradient Boosting).


## Before we beigin, let's cofigure some stuff so that the notebook runs both on your local machine and on *Google's Colaboratory*

1- If you are running locally, run the following cell:

In [ ]:
proj_dir = "proj-dir/"

2- If you are running on *Colab*, 
- Make sure you have uploaded all the project files to your *Google Drive*. Then, mount your drive by running the following cell:

In [1]:
from google.colab import drive
drive.mount("/content/drive")

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


- Then write out the path to the project files relative to your drive's root directory after `/content/drive/My Drive/` in the following cell:

In [1]:
proj_dir = "/content/drive/My Drive/Projects/human-activity-recognizer/" + "proj-dir/"

## Next up, let's import everything we need. Run the following:

If you don't have XGBoost installed, run the following:

In [4]:
!pip install --upgrade xgboost

     |████████████████████████████████| 127.6MB 96kB/s 
  Found existing installation: xgboost 0.90
    Uninstalling xgboost-0.90:
      Successfully uninstalled xgboost-0.90


In [2]:
import pandas as pd
from tensorflow.keras import models
from tensorflow.keras import layers
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn import datasets
from sklearn.metrics import accuracy_score
import xgboost as xgb
import numpy as np
import joblib

## We will solve this problem using different models. We will start by using simple Decision Trees and then go about trying Support Vector Machines, then finally, XGBoost:

Let's first have a quick review of how the sklearn API works. We will do this by using the Iris dataset.

In [6]:
data = datasets.load_iris()

In [7]:
data.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [8]:
print(data['DESCR'])

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [9]:
X,y = data["data"], data["target"]

In [10]:
X.shape

(150, 4)

In [11]:
y.shape

(150,)

In [12]:
clf = DecisionTreeClassifier()

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2)

In [14]:
X_train.shape

(120, 4)

In [15]:
clf.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [16]:
clf.score(X_test, y_test)

0.9666666666666667

## We got a 96% accuracy! Now let's try XGBoost:

In [17]:
Dtrain = xgb.DMatrix(X_train, y_train)

In [18]:
Dtrain

In [19]:
Dtest = xgb.DMatrix(X_test, y_test)

In [20]:
param = {
    'eta': 0.3, 
    'max_depth': 3,  
    'objective': 'multi:softprob',  
    'num_class': 3}

In [21]:
steps = 20

In [22]:
model = xgb.train(param, Dtrain, steps)

In [23]:
preds = model.predict(Dtest)

In [24]:
preds.shape

(30, 3)

In [25]:
best_preds = np.asarray([np.argmax(line) for line in preds])

In [26]:
best_preds

array([2, 0, 2, 2, 0, 1, 1, 2, 0, 2, 0, 0, 0, 0, 2, 0, 1, 2, 1, 1, 1, 1,
       2, 2, 1, 0, 1, 1, 0, 2])

In [27]:
best_preds.shape

(30,)

In [28]:
print(f"Accuracy = {accuracy_score(y_test, best_preds)}")

Accuracy = 1.0


## That's a 100 percent! Okay, now let's get back to our original dataset:

In [3]:
data = pd.read_csv(proj_dir+"X_train.txt", header=None, sep=" ")

In [4]:
data

0         1         2         3    ...       558       559       560       561
0     NaN  0.288585 -0.020294 -0.132905  ... -0.018446 -0.841247  0.179941 -0.058627
1     NaN  0.278419 -0.016411 -0.123520  ...  0.703511 -0.844788  0.180289 -0.054317
2     NaN  0.279653 -0.019467 -0.113462  ...  0.808529 -0.848933  0.180637 -0.049118
3     NaN  0.279174 -0.026201 -0.123283  ... -0.485366 -0.848649  0.181935 -0.047663
4     NaN  0.276629 -0.016570 -0.115362  ... -0.615971 -0.847865  0.185151 -0.043892
...   ...       ...       ...       ...  ...       ...       ...       ...       ...
7347  NaN  0.299665 -0.057193 -0.181233  ... -0.425619 -0.791883  0.238604  0.049819
7348  NaN  0.273853 -0.007749 -0.147468  ...  0.400219 -0.771840  0.252676  0.050053
7349  NaN  0.273387 -0.017011 -0.045022  ...  0.701169 -0.779133  0.249145  0.040811
7350  NaN  0.289654 -0.018843 -0.158281  ... -0.589479 -0.785181  0.246432  0.025339
7351  NaN  0.351503 -0.012423 -0.203867  ... -0.616956 -0.783267  0.246809  0.036695

[7352 rows x 562 columns]

## Doing some data cleaning:

In [5]:
with open(proj_dir+"features.txt") as file:
    columns = file.readlines()

In [6]:
df = data.dropna(axis=1)

In [7]:
df

1         2         3    ...       559       560       561
0     0.288585 -0.020294 -0.132905  ... -0.841247  0.179941 -0.058627
1     0.278419 -0.016411 -0.123520  ... -0.844788  0.180289 -0.054317
2     0.279653 -0.019467 -0.113462  ... -0.848933  0.180637 -0.049118
3     0.279174 -0.026201 -0.123283  ... -0.848649  0.181935 -0.047663
4     0.276629 -0.016570 -0.115362  ... -0.847865  0.185151 -0.043892
...        ...       ...       ...  ...       ...       ...       ...
7347  0.299665 -0.057193 -0.181233  ... -0.791883  0.238604  0.049819
7348  0.273853 -0.007749 -0.147468  ... -0.771840  0.252676  0.050053
7349  0.273387 -0.017011 -0.045022  ... -0.779133  0.249145  0.040811
7350  0.289654 -0.018843 -0.158281  ... -0.785181  0.246432  0.025339
7351  0.351503 -0.012423 -0.203867  ... -0.783267  0.246809  0.036695

[7352 rows x 561 columns]

In [8]:
cols = []
for c in columns:
    cols.append(c.strip("\n"))

In [9]:
df.columns = cols

In [10]:
labels = pd.read_csv(proj_dir+"y_train.txt")

In [11]:
labels = labels.to_numpy()

In [12]:
x_train = df.to_numpy()

In [13]:
y_train = labels

## Now let's define the model:

In [66]:
model = DecisionTreeClassifier()

In [68]:
model.fit(x_train[:-1], np.squeeze(y_train))

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [69]:
x_train.shape

(7352, 561)

In [70]:
y_train.shape

(7351, 1)

## Now let's evaluate it:

In [14]:
data = pd.read_csv(proj_dir+"X_test.txt", header=None, sep=" ")

In [15]:
data

0         1         2         3    ...       558       559       560       561
0     NaN  0.257178 -0.023285 -0.014654  ...  0.271151 -0.720009  0.276801 -0.057978
1     NaN  0.286027 -0.013163 -0.119083  ...  0.920593 -0.698091  0.281343 -0.083898
2     NaN  0.275485 -0.026050 -0.118152  ...  0.145068 -0.702771  0.280083 -0.079346
3     NaN  0.270298 -0.032614 -0.117520  ...  0.296407 -0.698954  0.284114 -0.077108
4     NaN  0.274833 -0.027848 -0.129527  ... -0.118545 -0.692245  0.290722 -0.073857
...   ...       ...       ...       ...  ...       ...       ...       ...       ...
2942  NaN  0.310155 -0.053391 -0.099109  ... -0.698885 -0.651732  0.274627  0.184784
2943  NaN  0.363385 -0.039214 -0.105915  ...  0.322549 -0.655181  0.273578  0.182412
2944  NaN  0.349966  0.030077 -0.115788  ...  0.363139 -0.655357  0.274479  0.181184
2945  NaN  0.237594  0.018467 -0.096499  ... -0.008398 -0.659719  0.264782  0.187563
2946  NaN  0.153627 -0.018437 -0.137018  ... -0.024965 -0.660080  0.263936  0.188103

[2947 rows x 562 columns]

In [16]:
labels = pd.read_csv(proj_dir+"y_test.txt")

In [17]:
labels = labels.to_numpy()

In [18]:
x_test = data.dropna(axis=1).to_numpy()

In [19]:
y_test = labels

In [20]:
x_test = x_test[:-1]

In [21]:
x_test.shape

(2946, 561)

In [22]:
y_test.shape

(2946, 1)

In [24]:
joblib.dump(model,proj_dir+"./trained_tree.sav")

In [81]:
model.score(x_test, np.squeeze(y_test))

0.7824168363883232

## 78% accuracy, fair enough. But we can do better...

## Let's try SVC:

In [82]:
model = SVC()

In [83]:
model.fit(x_train[:-1], np.squeeze(y_train))

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [84]:
model.score(x_test, np.squeeze(y_test))

0.9164969450101833

In [85]:
joblib.dump(model,proj_dir+"./trained_svc.sav")

['/content/drive/My Drive/Projects/human-activity-recognizer/proj-dir/./trained_svc.sav']

## Wow, we got a 91.6% accuracy! But let's see what XGBoost has to offer

In [25]:
y_train = y_train.squeeze()
y_test = y_test.squeeze()
y_train -= 1
y_test -= 1

In [26]:
## Let's remind us how many classes we have:
pd.Series(y_train).value_counts()

5    1407
4    1373
3    1286
0    1226
1    1073
2     986
dtype: int64

In [37]:
param = {
    'eta': 0.3, 
    'max_depth': 5,  
    'objective': 'multi:softprob',  
    'num_class': 6,
    }

In [38]:
steps = 40

In [39]:
Dtrain = xgb.DMatrix(x_train[:-1],y_train)

In [40]:
Dtest = xgb.DMatrix(x_test, y_test.squeeze())

In [41]:
model = xgb.train(param, Dtrain, steps)

In [42]:
preds = model.predict(Dtest)

In [43]:
best_preds = np.asarray([np.argmax(line) for line in preds])

In [44]:
best_preds

array([4, 4, 4, ..., 1, 1, 1])

In [45]:
best_preds.shape

(2946,)

In [46]:
print(f"Accuracy = {accuracy_score(y_test, best_preds)}")

Accuracy = 0.9059742023082146


## Thus SVMs were better suited for this task

# The End?

## 👀 Contact

If you want to contact me you can reach me at <zenyc@live.com>.